package/MDAnalysis/topology/MOL2Parser.py

```diff
@@ -148,9 +150,34 @@ def parse(self, **kwargs):
         resids = []
         resnames = []
         charges = []
-
+        set_charge = True
+        for line in sections['molecule']:
+            if line.strip().lower() == 'no_charges':
+                set_charge = False
+                break
         for a in atom_lines:
-            aid, name, x, y, z, atom_type, resid, resname, charge = a.split()[:9]
+            columns = a.split()
+            if len(columns) >= 9:
+                columns = columns[:9]
+                aid, name, x, y, z, atom_type, resid, resname, charge = columns
+            elif len(columns) < 6:
+                raise ValueError(f"The @<TRIPOS>ATOM block in mol2 file"
+                                 f" {os.path.basename(self.filename)}"
+                                 f" should contain at least 6 fields to be"
+                                 f" unpacked: atom_id atom_name x y z"
+                                 f" atom_type [subst_id[subst_name"
+                                 f" [charge [status_bit]]]]")
+            else:
+                aid, name, x, y, z, atom_type = columns[:6]
+                resid = '1'
+                resname = ''
+                charge = ''
+                opt_values = [resid, resname, charge]
+                for i in range(6, len(columns)):
+                    opt_values[i-6] = columns[i]
+                resid, resname, charge = opt_values
+                if len(charge):
+                    set_charge = False
```

Comment:
I think this is the opposite of what you wanted to achieve? `len(charge)` is `False` when a charge is missing (i.e. when it remains the empty string that you set as default), which is when you would want to set `set_charge` to `False`? In any case, this is untested...

BTW, you can also evaluate the string directly (`if charge:` or `if not charge`:).

From the file format specification it is unclear to me if this is allowed at all, but I'd think it isn't since the format explicitly specify if charges are present or not. If this is the case (probably someone else knows more) you can just skip charges in this whole block.

Reply:
@RMeli makes a good point. If the file format requires at the beginning of the file to specify if charges are present then we should just rely on this information. @Mrqeoqqt please read the format specification and let us know what should happen in the following cases:

1. MOLECULE section contains NO_CHARGES  as the *charge_type*
   1.  and **no charges** are present in ATOM section
   2. but **charges are present** in ATOM section
2. MOLECULE section does **not** contain NO_CHARGES (i.e., it contains one of the other keywords indicating some sort of charges, see manual)
   1.  and  **charges are present** in ATOM section
   2. but **no charges** are present in ATOM section

Obviously, 1(i) and 2(i) are the expected cases and should just work without any errors or warnings.

Case 2(ii) should definitely raise an exception.

For Case 1(ii) I am not sure: I feel it should  raise an exception (because the content of the file is inconsistent). However, one could also decide to ignore the charges (and issue a warning) or to read the charges nevertheless (I agree with @RMeli that we should not do that, i.e., skip this). @Mrqeoqqt what is your opinion here?

Reply:
There are multiple charge types (see below). I suppose except from "NO_CHARGE", the other keywords should indicate charges need to be provided. But the document didn't mention how to deal with case1(ii) and case 2(ii) if happens.
> NO_CHARGES, DEL_RE, GASTEIGER, GAST_HUCK, HUCKEL, 
> PULLMAN, GAUSS80_CHARGES, AMPAC_CHARGES, 
> MULLIKEN_CHARGES, DICT_ CHARGES, MMFF94_CHARGES, 
> USER_CHARGES

@orbeckst What about I implement the situations you have described and deal with warnings and errors?

I also see an interesting [topic](https://sourceforge.net/p/openbabel/mailman/message/31552037/) for `obabel` when trying to add MMFF94 charge model to `MOL2`: no matter what charge model is used, `obabel` uses `USER_CHARGES` in output.

Reply:
Case 2(ii) should just raise an exception (`ValueError`).

If there's nothing defined for Case 1(ii) then I would say we also raise a `ValueError` — that's safer than trying to guess intent and interpreting what is essentially a non-conformant file.

This should simplify your code because you can use the analysis of NO_CHARGES to decide how to handle the rest.